slides/week2/week2.org

```diff
@@ -1,133 +1,188 @@
+- [[http://cpp.sh/4mbgg]]
+
+* Includes
+- Why do we need =#include <iostream>=?
+ - [[http://cpp.sh/7jb5t]]
+- Includes let you bring other code into a file
+- Use this for
+ - Standard library functions and data types (anything std::)
+ - Other installed libraries (e.g. ROS, OpenCV, Qt)
+ - Splitting up a large program into multiple files
+- Without including anything, C++ is very limited
+
+* Include syntax
+- =<>= gets code from standard library or apt-installed libraries
+- =""= gets code from a nearby folder
+- If =""= does not find the file, it reverts to =<>= behavior
+- What's in the =<>= or =""= is a file name
+ - Traditionally the file name has extension ".h" or ".hpp"
+- Examples:
+#+BEGIN_SRC C++ -n
+#include <string>  // standard library
+#include <QWidget>  // file from Qt installed through apt
+#include <ros/ros.h>  // file from ros installed through apt
+#include "include/my_interface.hpp"  // another file in the same project,
+                                     // in a folder called "include"
+#+END_SRC
+
+* Namespaces
+- What happens when two things are assigned the same name?
+ - [[http://cpp.sh/24v43]]
+- Solution: protect your variables and function names with a unique namespace
+ - [[http://cpp.sh/66mrj]]
+ - Use keywords from a namespace using the =::= operator
+  - types: =std::string=, =my_library::MyClass=
+  - functions: =std::min_element=
+  - static variables: =std::string::npos=
+- Everything that is included should be in a namespace
+
+* "using namespace" keyword
+- [[http://cpp.sh/4d5gz]]
+- Handy (used in RJ) but also defeats the purpose of namespaces
+- Use with caution, only if each of the following are true:
+ - You use this namespace repeatedly in your code
+ - It is already clear without the namespace where the function comes from
+
+* Strings
+- C string (still valid in C++):
+#+BEGIN_SRC C++ -n
+char[] s = "this is a string";
+#+END_SRC
+- C++ string:
+#+BEGIN_SRC C++ -n
+#include <string>
+std::string s1("this is a string");  // constructor
+std::string s2 = "this is a string";  // same effect as constructor
+#+END_SRC
+- Unlike in C, C++ strings are a class instances and have methods
+
+* String Methods
+- What methods can I use for a string? (there are lots)
+ - Google! [[https://www.google.com/search?q=c%2B%2B+std%3A%3Astring]]
+- Adding to end:  + and += operators, =append(string)=, =push_back(char)=
+ - [[http://cpp.sh/84u76]]
+- Reading user input:  =std::cin >>=
+ - [[http://cpp.sh/64mc]]
+- Access characters like an array
+ - [[http://cpp.sh/9gb4x]]
+- =size()= and =length()= each get number of characters
+
+* Excercise
+- Find a buddy (or work alone if you want)
+- Write a function =make_palindrome=
+ - Input: string
+ - Output: string with reversed copy attached
+ - =make_palindrome("apple")= returns ="appleelppa"=
+- Starter code: [[http://cpp.sh/844tx]]
+
+* Solution
+- Using what we've learned so far
+ - [[http://cpp.sh/92y54]]
+- Preview for next week
```

Comment:
Might want to move to the next week. So the next week starts out as hey look what we did, see how much easier it is with the STL.